File: backend/app/models/alert.py
from sqlalchemy import Column, Integer, String, Float, Boolean, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from datetime import datetime
from app.models.base import Base

class Alert(Base):
    """Alerts created by users for notifiable sound classes"""
    __tablename__ = "alerts"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    class_id = Column(Integer, ForeignKey("notifiable_classes.id"), nullable=False)
    latitude = Column(Float, nullable=False)
    longitude = Column(Float, nullable=False)
    confidence = Column(Float, nullable=False)
    device_id = Column(String(255), nullable=False)
    is_verified = Column(Boolean, default=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    expires_at = Column(DateTime, nullable=True)  # When this alert should expire
    
    # Relationships
    user = relationship("User", back_populates="alerts")
    alert_class = relationship("NotifiableClass", back_populates="alerts")

File: mobile/lib/models/alert.dart
import 'notifiable_class.dart';

class Alert {
  final int id;
  final int classId;
  final double latitude;
  final double longitude;
  final double confidence;
  final String deviceId;
  final bool isVerified;
  final DateTime createdAt;
  final NotifiableClass alertClass;
  final double? distanceKm;

  Alert({
    required this.id,
    required this.classId,
    required this.latitude,
    required this.longitude,
    required this.confidence,
    required this.deviceId,
    required this.isVerified,
    required this.createdAt,
    required this.alertClass,
    this.distanceKm,
  });

  factory Alert.fromJson(Map<String, dynamic> json) {
    return Alert(
      id: json['id'],
      classId: json['class_id'],
      latitude: json['latitude'].toDouble(),
      longitude: json['longitude'].toDouble(),
      confidence: json['confidence'].toDouble(),
      deviceId: json['device_id'],
      isVerified: json['is_verified'],
      createdAt: DateTime.parse(json['created_at']),
      alertClass: NotifiableClass.fromJson(json['alert_class']),
      distanceKm: json['distance_km']?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'class_id': classId,
      'latitude': latitude,
      'longitude': longitude,
      'confidence': confidence,
      'device_id': deviceId,
      'is_verified': isVerified,
      'created_at': createdAt.toIso8601String(),
      'alert_class': alertClass.toJson(),
      'distance_km': distanceKm,
    };
  }
}